src/referenced-fields.ts

```diff
@@ -294,3 +332,256 @@ export const getModifiedFields: EngineInstance['getModifiedFields'] = (
+		case 'Select':
+			const selectNode = abstractSql as SelectNode;
+			for (const arg of selectNode[1]) {
+				assertAbstractSqlIsNotLegacy(arg);
+				sum += countTableSelects(arg, table);
+			}
+
+			return sum;
+
+		// Uninteresting atomic nodes
+		case 'Boolean':
+		case 'Date':
+		case 'Duration':
+		case 'EmbeddedText':
+		case 'GroupBy':
+		case 'Integer':
+		case 'JSON':
+		case 'Null':
+		case 'Number':
+		case 'ReferencedField':
+		case 'Text':
+			return 0;
+
+		// The atomic node we're looking for: a table selection
+		case 'Table':
+			const tableNode = abstractSql as TableNode;
+
+			if (tableNode[1] === table) {
+				return 1;
+			} else {
+				return 0;
+			}
+
+		default:
+			throw new Error(`unknown abstract sql type: ${abstractSql[0]}`);
+	}
+};
+
+// TODO:
+// - This function only narrows the root table of the rule. This is always
+//   safe when the root table isn't selected from more than once and it is not
+//   negated in the LF. Right now we conservatively check for the former but
+//   not the second. The negative forms (e.g. it is forbidden that ...) are
+//   not fully supported anyway.
+// - Removing multiple candidates selecting from the same database table to
+//   avoid visibility issues is too conservative. The correct criteria is to
+//   just remove any that are present in at least 2 disjoint subqueries.
+//   Because in this case the problem is that in those cases there is not a
+//   single place in the query that has visibility inside both disjoint
+//   subqueries and that is a requirement for adding the corrent binds for
+//   narrowing.
+// - We assume the ID column is named "id".
+// - This is a very restricted implementation of narrowing which could be
+//   expanded to cover more situations.
+//
+// This function modifies `abstractSql` in place.
+export const insertAffectedIdsBinds = (
+	abstractSql: AbstractSqlQuery,
+	lfRuleInfo: LfRuleInfo,
+) => {
+	const rootTableSelectCount = countTableSelects(
+		abstractSql,
+		lfRuleInfo.root.table,
+	);
+	if (rootTableSelectCount !== 1) {
+		return;
+	}
+
+	const narrowing: OrNode = [
+		'Or',
+		['Equals', ['Bind', lfRuleInfo.root.table], ['EmbeddedText', '{}']],
+		[
+			'Equals',
+			['ReferencedField', lfRuleInfo.root.alias, 'id'],
+			['Any', ['Bind', lfRuleInfo.root.table], 'Integer'],
+		],
+	];
+
+	// Assume (but check) that the query is of the form:
+	//
+	// SELECT (SELECT COUNT(*) ...) = 0
+	if (
+		abstractSql[0] !== 'Equals' ||
+		abstractSql[1][0] !== 'SelectQuery' ||
+		abstractSql[2][0] !== 'Number'
+	) {
+		throw new Error(
+			'Query is not of the form: SELECT (SELECT COUNT(*) ...) = 0',
+		);
+	}
+
+	const selectQueryNode = abstractSql[1] as SelectQueryNode;
+	const whereStatement = selectQueryNode.slice(1).find((x) => x[0] === 'Where');
+	if (whereStatement === undefined) {
+		selectQueryNode.push(['Where', narrowing]);
+	} else {
+		const whereNode = whereStatement as WhereNode;
+		whereNode[1] = ['And', whereNode[1], narrowing];
+	}
```

Comment:
I assume there's an `isWhereNode` (if not it makes sense to create imo):
```suggestion
	const whereNode = selectQueryNode.slice(1).find(isWhereNode);
	if (whereNode === undefined) {
		selectQueryNode.push(['Where', narrowing]);
	} else {
		whereNode[1] = ['And', whereNode[1], narrowing];
	}
```